File: code/backend/python/app/adapters/rest/story/story_admin_controller.py
from fastapi import APIRouter, Query, Path, HTTPException, Body, Request
from typing import List, Dict, Any
from app.core.ports.story.story_query_port import StoryQueryPort
from app.core.ports.story.story_import_port import StoryImportPort
from app.core.ports.story.story_validator_port import StoryValidationException
from app.core.models.story.story_summary import StorySummary
from app.core.models.story.story_import_result import StoryImportResult

class StoryAdminController:
    def __init__(self, query_port: StoryQueryPort, import_port: StoryImportPort, validator_port=None):
        self.query_port = query_port
        self.import_port = import_port
        self.validator_port = validator_port
        self.router = APIRouter(prefix="/api/admin/stories", tags=["Story Admin"])

        self.router.add_api_route("", self.list_all_stories, methods=["GET"], response_model=List[StorySummary])
        self.router.add_api_route("/import", self.import_story, methods=["POST"], response_model=StoryImportResult, status_code=201)
        self.router.add_api_route("/{uuid}/validate", self.validate_story, methods=["GET"])
        self.router.add_api_route("/{uuid}", self.delete_story, methods=["DELETE"])

    async def list_all_stories(self, req: Request, lang: str = Query("en")) -> List[StorySummary]:
        self._require_admin(req)
        return self.query_port.list_all_stories(lang)

    async def import_story(self, req: Request, data: Dict[str, Any] = Body(default=None)) -> StoryImportResult:
        self._require_admin(req)
        if not data:
            raise HTTPException(status_code=400, detail={
                "error": "EMPTY_IMPORT_DATA",
                "message": "Request body must contain story data"
            })
        try:
            return self.import_port.import_story(data)
        except StoryValidationException as e:
            raise HTTPException(status_code=400, detail={
                "error": "INVALID_STORY",
                "message": e.report.summary(),
                "errors": [err.to_dict() for err in e.report.errors],
            })
        except ValueError as e:
            raise HTTPException(status_code=400, detail={
                "error": "INVALID_IMPORT_DATA",
                "message": str(e)
            })

    async def validate_story(self, req: Request, uuid: str = Path(...)):
        # Step 22: read-only integrity report for a persisted story.
        self._require_admin(req)
        if self.validator_port is None:
            return {"valid": True, "count": 0, "errors": []}
        report = self.validator_port.validate_story_by_uuid(uuid)
        if report is None:
            raise HTTPException(status_code=404, detail={
                "error": "STORY_NOT_FOUND",
                "message": f"No story found with UUID: {uuid}"
            })
        return report.to_dict()

    async def delete_story(self, req: Request, uuid: str = Path(...)):
        self._require_admin(req)
        deleted = self.import_port.delete_story(uuid)
        if not deleted:
            raise HTTPException(status_code=404, detail={
                "error": "STORY_NOT_FOUND",
                "message": f"No story found with UUID: {uuid}"
            })
        return {"status": "DELETED", "uuid": uuid}

    def _require_admin(self, request: Request):
        role = request.state.role if hasattr(request.state, "role") else None
        if role != "ADMIN":
            raise HTTPException(status_code=403, detail={
                "error": "FORBIDDEN",
                "message": "Insufficient permissions"
            })

File: code/backend/php/src/Adapter/Rest/Story/StoryAdminController.php
<?php

declare(strict_types=1);

namespace Games\Paths\Adapter\Rest\Story;

use Games\Paths\Core\Domain\Story\StoryValidationException;
use Games\Paths\Core\Port\Story\StoryImportPort;
use Games\Paths\Core\Port\Story\StoryQueryPort;
use Games\Paths\Core\Service\Story\StoryValidatorService;
use Psr\Http\Message\ResponseInterface as Response;
use Psr\Http\Message\ServerRequestInterface as Request;
use InvalidArgumentException;

class StoryAdminController
{
    private StoryQueryPort $queryPort;
    private StoryImportPort $importPort;
    private ?StoryValidatorService $validator;

    public function __construct(StoryQueryPort $queryPort, StoryImportPort $importPort, ?StoryValidatorService $validator = null)
    {
        $this->queryPort = $queryPort;
        $this->importPort = $importPort;
        $this->validator = $validator;
    }

    public function listAllStories(Request $request, Response $response, array $args): Response
    {
        if (!$this->isAdmin($request)) {
            return $this->forbidden($response);
        }

        $lang = $request->getQueryParams()['lang'] ?? 'en';
        $stories = $this->queryPort->listAllStories($lang);

        $response->getBody()->write(json_encode($stories));
        return $response->withHeader('Content-Type', 'application/json');
    }

    public function importStory(Request $request, Response $response, array $args): Response
    {
        if (!$this->isAdmin($request)) {
            return $this->forbidden($response);
        }

        $data = $request->getParsedBody();
        if (empty($data)) {
            $response->getBody()->write(json_encode([
                'error' => 'EMPTY_IMPORT_DATA',
                'message' => 'Request body must contain story data'
            ]));
            return $response->withStatus(400)->withHeader('Content-Type', 'application/json');
        }

        try {
            $result = $this->importPort->importStory($data);
            $response->getBody()->write(json_encode($result));
            return $response->withStatus(201)->withHeader('Content-Type', 'application/json');
        } catch (StoryValidationException $e) {
            $report = $e->getReport();
            $response->getBody()->write(json_encode([
                'error' => 'INVALID_STORY',
                'message' => $report->summary(),
                'errors' => array_map(fn ($err) => $err->toArray(), $report->getErrors()),
            ]));
            return $response->withStatus(400)->withHeader('Content-Type', 'application/json');
        } catch (InvalidArgumentException $e) {
            $response->getBody()->write(json_encode([
                'error' => 'INVALID_IMPORT_DATA',
                'message' => $e->getMessage()
            ]));
            return $response->withStatus(400)->withHeader('Content-Type', 'application/json');
        }
    }

    public function validateStory(Request $request, Response $response, array $args): Response
    {
        if (!$this->isAdmin($request)) {
            return $this->forbidden($response);
        }

        $uuid = $args['uuid'];
        if ($this->validator === null) {
            $response->getBody()->write(json_encode(['valid' => true, 'count' => 0, 'errors' => []]));
            return $response->withHeader('Content-Type', 'application/json');
        }
        $report = $this->validator->validateStoryByUuid($uuid);
        if ($report === null) {
            $response->getBody()->write(json_encode([
                'error' => 'STORY_NOT_FOUND',
                'message' => 'No story found with UUID: ' . $uuid
            ]));
            return $response->withStatus(404)->withHeader('Content-Type', 'application/json');
        }
        $response->getBody()->write(json_encode($report->toArray()));
        return $response->withHeader('Content-Type', 'application/json');
    }

    public function deleteStory(Request $request, Response $response, array $args): Response
    {
        if (!$this->isAdmin($request)) {
            return $this->forbidden($response);
        }

        $uuid = $args['uuid'];
        $deleted = $this->importPort->deleteStory($uuid);

        if (!$deleted) {
            $response->getBody()->write(json_encode([
                'error' => 'STORY_NOT_FOUND',
                'message' => 'No story found with UUID: ' . $uuid
            ]));
            return $response->withStatus(404)->withHeader('Content-Type', 'application/json');
        }

        $response->getBody()->write(json_encode([
            'status' => 'DELETED',
            'uuid' => $uuid
        ]));
        return $response->withHeader('Content-Type', 'application/json');
    }

    private function isAdmin(Request $request): bool
    {
        $role = $request->getAttribute('role');
        return $role === 'ADMIN';
    }

    private function forbidden(Response $response): Response
    {
        $response->getBody()->write(json_encode([
            'error' => 'FORBIDDEN',
            'message' => 'Insufficient permissions (ADMIN role required)'
        ]));
        return $response->withStatus(403)->withHeader('Content-Type', 'application/json');
    }
}
